finality-aleph/src/sync/message_limiter.rs

```diff
@@ -0,0 +1,181 @@
+use parity_scale_codec::Encode;
+
+use crate::sync::data::MAX_SYNC_MESSAGE_SIZE;
+
+pub const MSG_BYTES_LIMIT: usize = MAX_SYNC_MESSAGE_SIZE as usize;
+
+pub struct Limiter<'a, D: Encode, const LIMIT: usize> {
+    msg: &'a [D],
+    start_index: usize,
+    indexes: Vec<usize>,
+}
+
+pub type MsgLimiter<'a, D> = Limiter<'a, D, MSG_BYTES_LIMIT>;
+
+impl<'a, D: Encode, const LIMIT: usize> Limiter<'a, D, LIMIT> {
+    pub fn new(msg: &'a [D]) -> Self {
+        Self {
+            msg,
+            start_index: 0,
+            indexes: (0..=msg.len()).collect(),
+        }
+    }
+
+    fn next_largest_msg(&mut self) -> Option<&'a [D]> {
+        if self.start_index == self.msg.len() {
+            return None;
+        }
+        let end_idx = self.bs_by_encode()?;
+
+        let start_index = self.start_index;
+        self.start_index = end_idx;
+
+        Some(&self.msg[start_index..end_idx])
+    }
+
+    fn bs_by_encode(&self) -> Option<usize> {
+        let indexes = &self.indexes[self.start_index..];
+
+        let idx = indexes.partition_point(|&idx| {
+            let encoded_size = self.msg[self.start_index..idx].encoded_size();
+            encoded_size <= LIMIT
+        }) - 1; // minus 1 since this is the first index where encoded size is larger than LIMIT
+
+        let idx = idx + self.start_index;
+
+        if idx == self.start_index {
+            None
```

Comment:
this means that a single item is larger than the message size limit, right? if so, this should definitely be a critical error

Reply:
imho we could even just panic here as it would be really bad if one item is larger than the message size limit as you have pointed it out

Reply:
I'd say we return an error, catch and display it somewhere higher as "critical", and hope that this particular block won't get finalized and we'll prune it soon...

Reply:
ok, done